vllm/model_executor/models/llama_with_lora.py

```diff
@@ -0,0 +1,424 @@
+            self.dropout = nn.Dropout(dropout)
+        else:
+            self.dropout = nn.Identity()
+
+    @classmethod
+    def from_state_dict(cls, state_dict):
+        expected_keys = set(['loras', 'dropout', 'scaling'])
+        if state_dict.keys() != expected_keys:
+            raise ValueError(f"Expected keys {expected_keys}, got {state_dict.keys()}")
+
+        rank, hidden_dim = state_dict['loras']['lora_A.weight'].shape
+        dropout = state_dict['dropout']
+        scaling = state_dict['scaling']
+        lora_layer = cls(hidden_dim=hidden_dim, rank=rank, dropout=dropout, scaling=scaling)
+        lora_layer.load_state_dict(state_dict['loras'])
+        return lora_layer
+
+    def forward(self, x):
+        return self.lora_B(self.lora_A(self.dropout(x))) * self.scaling
+
+
+class LlamaAttention(nn.Module):
+
+    def __init__(
+        self,
+        hidden_size: int,
+        num_heads: int,
+        num_kv_heads: int,
+        rope_theta: float = 10000,
+    ):
+        super().__init__()
+        self.hidden_size = hidden_size
+        tp_size = get_tensor_model_parallel_world_size()
+        self.total_num_heads = num_heads
+        assert self.total_num_heads % tp_size == 0
+        self.num_heads = self.total_num_heads // tp_size
+        self.total_num_kv_heads = num_kv_heads
+        assert self.total_num_kv_heads % tp_size == 0
+        self.num_kv_heads = self.total_num_kv_heads // tp_size
+        self.head_dim = hidden_size // self.total_num_heads
+        self.q_size = self.num_heads * self.head_dim
+        self.kv_size = self.num_kv_heads * self.head_dim
+        self.scaling = self.head_dim**-0.5
+        self.rope_theta = rope_theta
+
+        self.qkv_proj = ColumnParallelLinear(
+            hidden_size,
+            (self.total_num_heads + 2 * self.total_num_kv_heads) *
+            self.head_dim,
+            bias=False,
+            gather_output=False,
+            perform_initialization=False,
+        )
+        self.o_proj = RowParallelLinear(
+            self.total_num_heads * self.head_dim,
+            hidden_size,
+            bias=False,
+            input_is_parallel=True,
+            perform_initialization=False,
+        )
+        self.attn = PagedAttentionWithRoPE(self.num_heads,
+                                           self.head_dim,
+                                           self.scaling,
+                                           base=self.rope_theta,
+                                           rotary_dim=self.head_dim,
+                                           num_kv_heads=self.num_kv_heads)
+
+        self.q_lora = None
+        self.v_lora = None
+
+    def load_lora(self, q_lora_state_dict, v_lora_state_dict):
+        # TODO (Moin): generalize this abstraction later on and allow flexible devices
+        self.q_lora = LinearLoRALayer.from_state_dict(q_lora_state_dict).to("cuda")
+        self.v_lora = LinearLoRALayer.from_state_dict(v_lora_state_dict).to("cuda")
```

Comment:
it might be helpful to do load_lora that just returns LinearLoRALayer vs set_lora that sets self.q_lora/self.v_lora -- so that you can return to loras that were loaded in the past